backend/http/http.go

```diff
@@ -0,0 +1,345 @@
+	"strings"
+	"time"
+
+	"github.com/go-git/go-billy/v5/osfs"
+	"github.com/go-git/go-git/v6/plumbing/storer"
+	"github.com/go-git/go-git/v6/plumbing/transport"
+	"github.com/go-git/go-git/v6/storage"
+	"github.com/go-git/go-git/v6/utils/ioutil"
+)
+
+type contextKey string
+
+type service struct {
+	pattern  *regexp.Regexp
+	method  string
+	handler http.HandlerFunc
+	svc     transport.Service
+}
+
+var services = []service{
+	{regexp.MustCompile("(.*?)/HEAD$"), http.MethodGet, getTextFile, ""},
+	{regexp.MustCompile("(.*?)/info/refs$"), http.MethodGet, getInfoRefs, ""},
+	{regexp.MustCompile("(.*?)/objects/info/alternates$"), http.MethodGet, getTextFile, ""},
+	{regexp.MustCompile("(.*?)/objects/info/http-alternates$"), http.MethodGet, getTextFile, ""},
+	{regexp.MustCompile("(.*?)/objects/info/packs$"), http.MethodGet, getInfoPacks, ""},
+	{regexp.MustCompile("(.*?)/objects/[0-9a-f]{2}/[0-9a-f]{38}$"), http.MethodGet, getLooseObject, ""},
+	{regexp.MustCompile("(.*?)/objects/[0-9a-f]{2}/[0-9a-f]{62}$"), http.MethodGet, getLooseObject, ""},
+	{regexp.MustCompile("(.*?)/objects/pack/pack-[0-9a-f]{40}\\.pack$"), http.MethodGet, getPackFile, ""},
+	{regexp.MustCompile("(.*?)/objects/pack/pack-[0-9a-f]{64}\\.pack$"), http.MethodGet, getPackFile, ""},
+	{regexp.MustCompile("(.*?)/objects/pack/pack-[0-9a-f]{40}\\.idx$"), http.MethodGet, getIdxFile, ""},
+	{regexp.MustCompile("(.*?)/objects/pack/pack-[0-9a-f]{64}\\.idx$"), http.MethodGet, getIdxFile, ""},
+
+	// TODO: Support git-upload-archive
+	// {regexp.MustCompile("(.*?)/git-upload-archive$"), http.MethodPost, serviceRpc, transport.UploadArchiveService},
+	{regexp.MustCompile("(.*?)/git-upload-pack$"), http.MethodPost, serviceRpc, transport.UploadPackService},
+	{regexp.MustCompile("(.*?)/git-receive-pack$"), http.MethodPost, serviceRpc, transport.ReceivePackService},
+}
+
+// DefaultLoader is the default loader used to load repositories from storage.
+// It will use the current working directory as the base path for the
+// repositories.
+var DefaultLoader = transport.NewFilesystemLoader(osfs.New("."), false)
+
+// Handler is a Git HTTP handler that serves git repositories over HTTP.
+//
+// It supports serving repositories using both the Smart-HTTP and the Dumb-HTTP
+// protocols. When the Dumb-HTTP protocol is used, the repository store must
+// implement the [storer.FilesystemStorer] interface. Keep in mind that
+// repositories that wish to be server using the Dumb-HTTP protocol must update
+// the server info files. This can be done by using
+// [transport.UpdateServerInfo] before serving the repository.
+type Handler struct {
+	// Loader is the loader used to load repositories from storage. If nil, the
+	// default loader is used.
+	Loader transport.Loader
+	// ErrorLog is the logger used to log errors. If nil, no errors are logged.
+	ErrorLog *log.Logger
+	// Prefix is a path prefix that will be stripped from the URL path before
+	// matching the service patterns.
+	Prefix string
```

Comment:
What if I want postfix, maybe make a function?

Reply:
Thinking now, it can be resolved in the loader.

Reply:
You can't have postfix because that will interfere with git routes ex. given `POST /my/long/prefix/gitrepo/git-receive-pack`, you can't have anything appended after `git-receive-pack` because that's what the git client expects the endpoint to be.

However, say you want to mount your middleware under `http://mydomain/gitstuff/` then you would use `Prefix: "/gitstuff"` and the middleware will strip that part when it tries to match for the repository name before giving it to the `Loader` to find.